CA2000 FP with ? or ?? operators on a return;

#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.3.0-beta2.final

#### Diagnostic ID

CA2000: Dispose objects before losing scope

#### Repro steps

Using VS2019 16.6.4, targetting .net framework 4.7.2

CA2000 seems to firing incorrectly on usages where a ?? or ? test is made, suggesting that it's branch analysis doesn't detect that the object is referenced in the branch, eg:
```
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace CA2000Test
{
    class Program
    {
        static void Main(string[] args)
        {
        }

        public IDisposable TestCa2000(string path, bool boolA, bool boolB)
        {
            if (boolA)
            {
                IDisposable entry = null;
                if (Dictionary.ContainsKey(path))
                {
                    entry = Dictionary[path];
                }

                return entry ?? new Process();
            }

            return !boolB ? new Process() : Dictionary[path];
        }

        public Dictionary<string, IDisposable> Dictionary { get; set; }
    }
}
```

The returns trigger a CA2000 warning, unless there's something very subtle it's trying to tell me, it appears to be a false positive.

#### Expected behavior
No warning

#### Actual behavior
Program.cs(26,29,26,42): warning CA2000: Call System.IDisposable.Dispose on object created by 'new Process()' before all references to it are out of scope.
Program.cs(23,33,23,46): warning CA2000: Call System.IDisposable.Dispose on object created by 'new Process()' before all references to it are out of scope.

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DisposeObjectsBeforeLosingScopeTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DisposeObjectsBeforeLosingScopeTests.cs
@@ -12327,5 +12327,72 @@ public class Class1
                 LanguageVersion = CSharpLanguageVersion.CSharp8,
             }.RunAsync();
         }
+
+        [Fact, WorkItem(3873, "https://github.com/dotnet/roslyn-analyzers/issues/3873")]
+        public async Task Dispose_ConditionalControlFlow_NoDiagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace CA2000Test
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+        }
+
+        public IDisposable TestCa2000(string path, bool boolA, bool boolB)
+        {
+            if (boolA)
+            {
+                IDisposable entry = null;
+                if (Dictionary.ContainsKey(path))
+                {
+                    entry = Dictionary[path];
+                }
+
+                return entry ?? new Process();
+            }
+
+            return !boolB ? new Process() : Dictionary[path];
+        }
+
+        public Dictionary<string, IDisposable> Dictionary { get; set; }
+    }
+}");
+        }
+
+        [Fact, WorkItem(3873, "https://github.com/dotnet/roslyn-analyzers/issues/3873")]
+        public async Task Dispose_ConditionalControlFlow_NoDiagnostic_02()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+class Program
+{
+    private IDisposable _field;
+    public Dictionary<string, IDisposable> Dictionary { get; set; }
+
+    public IDisposable TestCa2000(string path)
+    {
+        return Dictionary[path] ?? new Process();
+    }
+
+    public IDisposable TestCa2000()
+    {
+        return _field ?? new Process();
+    }
+
+    public IDisposable TestCa2000(IDisposable[] disposables)
+    {
+        return disposables[0] ?? new Process();
+    }
+}");
+        }
     }
 }

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/DefaultPointsToValueGenerator.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/DefaultPointsToValueGenerator.cs
@@ -35,16 +35,20 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis
                 {
                     return PointsToAbstractValue.NoLocation;
                 }
-                else if (analysisEntity.HasUnknownInstanceLocation ||
-                    PointsToAnalysisKind != PointsToAnalysisKind.Complete &&
-                    analysisEntity.IsChildOrInstanceMemberNeedingCompletePointsToAnalysis())
+                else if (analysisEntity.HasUnknownInstanceLocation)
                 {
                     return PointsToAbstractValue.Unknown;
                 }
 
                 value = PointsToAbstractValue.Create(AbstractLocation.CreateAnalysisEntityDefaultLocation(analysisEntity), mayBeNull: true);
-                _trackedEntitiesBuilder.AddEntityAndPointsToValue(analysisEntity, value);
-                _defaultPointsToValueMapBuilder.Add(analysisEntity, value);
+
+                // PERF: Do not track entity and its points to value for partial analysis for entities requiring complete analysis.
+                if (PointsToAnalysisKind == PointsToAnalysisKind.Complete ||
+                    !analysisEntity.IsChildOrInstanceMemberNeedingCompletePointsToAnalysis())
+                {
+                    _trackedEntitiesBuilder.AddEntityAndPointsToValue(analysisEntity, value);
+                    _defaultPointsToValueMapBuilder.Add(analysisEntity, value);
+                }
             }
 
             return value;

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs
@@ -81,6 +81,15 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis
 
             private bool ShouldBeTracked(AnalysisEntity analysisEntity)
                 => PointsToAnalysis.ShouldBeTracked(analysisEntity, DataFlowAnalysisContext.PointsToAnalysisKind);
+            private PointsToAbstractValue GetValueForEntityThatShouldNotBeTracked(AnalysisEntity analysisEntity)
+            {
+                Debug.Assert(!ShouldBeTracked(analysisEntity));
+                Debug.Assert(!CurrentAnalysisData.TryGetValue(analysisEntity, out var existingValue) || existingValue == PointsToAbstractValue.NoLocation);
+
+                return !PointsToAnalysis.ShouldBeTracked(analysisEntity.Type) ?
+                    PointsToAbstractValue.NoLocation :
+                    _defaultPointsToValueGenerator.GetOrCreateDefaultValue(analysisEntity);
+            }
 
             public override PointsToAnalysisData Flow(IOperation statement, BasicBlock block, PointsToAnalysisData input)
             {
@@ -141,8 +150,7 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis
             {
                 if (!ShouldBeTracked(analysisEntity))
                 {
-                    Debug.Assert(!CurrentAnalysisData.TryGetValue(analysisEntity, out var existingValue) || existingValue == PointsToAbstractValue.NoLocation);
-                    return !PointsToAnalysis.ShouldBeTracked(analysisEntity.Type) ? PointsToAbstractValue.NoLocation : PointsToAbstractValue.Unknown;
+                    return GetValueForEntityThatShouldNotBeTracked(analysisEntity);
                 }
 
                 if (!CurrentAnalysisData.TryGetValue(analysisEntity, out var value))
@@ -295,7 +303,7 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis
 
                 if (!ShouldBeTracked(analysisEntity))
                 {
-                    return PointsToAbstractValue.NoLocation;
+                    return GetValueForEntityThatShouldNotBeTracked(analysisEntity);
                 }
 
                 var location = AbstractLocation.CreateAllocationLocation(operation, analysisEntity.Type, DataFlowAnalysisContext);

